compiler/shared_ast/interpreter.ml

```diff
@@ -34,30 +34,88 @@ let is_empty_error : type a. (a, 'm) gexpr -> bool =
 let indent_str = ref ""
 
 (** {1 Evaluation} *)
-let print_log lang entry infos pos e =
-  if Global.options.trace then
-    match entry with
-    | VarDef _ ->
-      Message.log "%s%a %a: @{<green>%s@}" !indent_str Print.log_entry entry
-        Print.uid_list infos
-        (Message.unformat (fun ppf ->
-             (if Global.options.debug then Print.expr ~debug:true ()
-              else Print.UserFacing.expr lang)
-               ppf e))
-    | PosRecordIfTrueBool -> (
-      match pos <> Pos.no_pos, Mark.remove e with
-      | true, ELit (LBool true) ->
-        Message.log "%s@[<v>%a@{<green>Definition applied@}:@,%a@]" !indent_str
-          Print.log_entry entry Pos.format_loc_text pos
-      | _ -> ())
-    | BeginCall ->
-      Message.log "%s%a %a" !indent_str Print.log_entry entry Print.uid_list
-        infos;
-      indent_str := !indent_str ^ "  "
-    | EndCall ->
-      indent_str := String.sub !indent_str 0 (String.length !indent_str - 2);
-      Message.log "%s%a %a" !indent_str Print.log_entry entry Print.uid_list
-        infos
+
+let rec format_runtime_value lang ppf = function
+  | Runtime.Unit -> Print.UserFacing.unit lang ppf ()
+  | Runtime.Bool b -> Print.UserFacing.bool lang ppf b
+  | Runtime.Money m -> Print.UserFacing.money lang ppf m
+  | Runtime.Integer i -> Print.UserFacing.integer lang ppf i
+  | Runtime.Decimal d -> Print.UserFacing.decimal lang ppf d
+  | Runtime.Date t -> Print.UserFacing.date lang ppf t
+  | Runtime.Duration dt -> Print.UserFacing.duration lang ppf dt
+  | Runtime.Enum (name, (constr, v)) ->
+    Format.fprintf ppf "@[<hov 2>%s.%s@ (%a)@]" name constr
+      (format_runtime_value lang)
+      v
+  | Runtime.Struct (name, fields) ->
+    Format.fprintf ppf "@[<hv 2>%s {@ %a@;<1 -2>}@]" name
+      (Format.pp_print_list ~pp_sep:Format.pp_print_space (fun ppf (fld, v) ->
+           Format.fprintf ppf "@[<hov 2>-- %s:@ %a@]" fld
+             (format_runtime_value lang)
+             v))
+      fields
+  | Runtime.Array elts ->
+    Format.fprintf ppf "@[<hv 2>[@,@[<hov>%a@]@;<0 -2>]@]"
+      (Format.pp_print_list
+         ~pp_sep:(fun ppf () -> Format.fprintf ppf ";@ ")
+         (format_runtime_value lang))
+      (Array.to_list elts)
+  | Runtime.Unembeddable -> Format.pp_print_string ppf "<?>"
```

Comment:
If a user sees `<?>` in their trace logs, they will be very confused. A better print would be `<object>`, mirroring what e.g. Python does.